shopinvader_api_payment_provider_sips/routers/payment_sips.py

```diff
@@ -0,0 +1,118 @@
+import logging
+import pprint
+from typing import Annotated
+from urllib.parse import quote
+
+from fastapi import Depends, Request
+from fastapi.responses import RedirectResponse
+
+from odoo import api
+from odoo.exceptions import ValidationError
+
+from odoo.addons.fastapi.dependencies import odoo_env
+from odoo.addons.shopinvader_api_payment.routers import payment_router
+
+_logger = logging.getLogger(__name__)
+
+
+def _verify_sips_signature(tx_sudo, data):
+    """Verify the SIPS signature."""
+    expected_signature = tx_sudo.provider_id._sips_generate_shasign(data.get("Data"))
+    if not expected_signature:
+        _logger.warning("received sips notification without signature")
+        raise ValidationError()
+    if not hmac.compare_digest(data.get("Seal", ""), expected_signature):
+        _logger.warning("received sips notification with invalid signature")
+        raise ValidationError()
+
+
+def _tx_state_to_redirect_status(tx_state):
+    """Map a transaction state to a redirect status."""
+    if tx_state == "done":
+        return "success"
+    elif tx_state == "cancel":
+        return "cancelled"
+    else:
+        return "unknown"
+
+
+@payment_router.post("/payment/providers/sips/return")
+async def sips_return(
+    request: Request,
+    odoo_env: Annotated[api.Environment, Depends(odoo_env)],
+):
+    """Handle SIPS return.
+
+    After the payment, the user is redirected with a POST to this endpoint. We handle
+    the notification data to update the transaction status. We then redirect the browser
+    with a GET to the frontend_return_url, with the transaction reference as parameter.
+
+    Future: we could also return a unguessable transaction uuid that the front could the
+    use to consult /payment/transactions/{uuid} and obtain the transaction status.
+    """
+    data = await request.form()
+    _logger.info(
+        "return notification received from SIPS with data:\n%s", pprint.pformat(data)
+    )
+    # Check the integrity of the notification
+    tx_sudo = (
+        odoo_env["payment.transaction"]
+        .sudo()
+        ._get_tx_from_notification_data(
+            "sips",
+            data,
+        )
+    )
+    try:
+        _verify_sips_signature(tx_sudo, data)
+    except Exception:
+        return {"error": "Invalid sips signature"}
+
+    notification_data = tx_sudo._sips_notification_data_to_object(data.get("Data"))
+    returnContext = json.loads(notification_data.get("returnContext", "{}"))
+    reference = returnContext.get("reference")
+    frontend_redirect_url = returnContext.get("frontend_redirect_url")
+
+    try:
+        tx_sudo._handle_notification_data("sips", data)
+        status = _tx_state_to_redirect_status(tx_sudo.state)
+    except Exception:
+        _logger.exception("unable to handle sips notification data", exc_info=True)
+        status = "error"
+    return RedirectResponse(
+        url=f"{frontend_redirect_url}?status={status}&reference={quote(reference)}",
```

Comment:
We need a better way to construct the URL, in case the frontend as already provided query parameters in `frontend_redirect_urls`.

Reply:
Done in a new commit